Change Su-33 carrier flights to "Takeoff from runway"

MetalStormGhost opened the pull request:
Change the start_type for generated Su-33 flights from carriers to "Takeoff from runway" to work around a DCS AI issue preventing Su-33s from taking off when set to "Takeoff from ramp".

I ran into the issue reported earlier (#1352) in a coop campaign with a friend last night and I can confirm the problem. I made a really simple change to try and work around it. I did test that it generates the correct flights by opening the generated mission in the mission editor, but didn't test it by actually running the mission yet (will do that today).

Also, I wanted to create a PR for this to get advice how to properly access game.settings (or the equivalent from the settings rework that might be happening in the background) from within packagebuilder. I was thinking of making this an option, since I'm hopeful the actual DCS issue might get fixed at some point and we wouldn't need this after that anymore.

MetalStormGhost committed:
Change the start_type for generated Su-33 flights from carriers to "Takeoff from runway" to work around a DCS AI issue preventing Su-33s from taking off when set to "Takeoff from ramp". Possibly resolves #1352

DanAlbert commented:
> I was thinking of making this an option, since I'm hopeful the actual DCS issue might get fixed at some point and we wouldn't need this after that anymore.

When that day comes we can just remove the workaround. I don't think an option is needed here, since the option is effectively "do you want broken AI Su-33s or not?". No one does, so no option needed :)

DanAlbert reviewed (approved):
I see this is still a draft, but LGTM. Will avoid merging before you undraft it though.

MetalStormGhost commented:
Yeah, I read through the issue one more time and realized this problem probably only applies to the free Kuznetsov. I'll change it so the supercarrier start types are not affected.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into dcs-issue-su33-ai-carrier-takeoff

MetalStormGhost committed:
Added a check for "KUZNECOW" carrier type in changing the start_type for generated Su-33 flights from carriers to "Takeoff from runway".

MetalStormGhost committed:
Merge branch 'dcs-issue-su33-ai-carrier-takeoff' of https://github.com/MetalStormGhost/dcs_liberation into dcs-issue-su33-ai-carrier-takeoff

MetalStormGhost committed:
Added a check for "KUZNECOW" carrier type in changing the start_type for generated Su-33 flights from carriers to "Takeoff from runway".

MetalStormGhost committed:
Re-committing my latest changes after fixing a broken black integration in the IDE.

MetalStormGhost commented:
I think I need some advice from someone with a bit more insight on the inner workings of Liberation. I tried checking if the carrier is a free Kuznetsov (as opposed to the supercarrier one), but I just realized that upgrade_to_supercarrier is run after packagebuilder does its thing.

Should the whole thing be moved away from packagebuilder to a stage that is run upon mission generation?

MetalStormGhost committed:
Reverted my changes to packagebuilder.py and reimplemented the checking for Su-33 and Kuznetsov in aircraft.py, which is run on mission generation instead of mission completion.

MetalStormGhost committed:
Corrected indentation in _start_type_at_group parameters.

MetalStormGhost commented:
I moved the logic from packagebuilder to aircraft.py and it seems to work based on my initial testing. I'm removing the draft status.

DanAlbert commented on gen/aircraft.py:
```diff
@@ -265,6 +266,25 @@ def _start_type(start_type: str) -> StartType:
+            return StartType.Runway
+        elif start_type == "Cold":
+            return StartType.Cold
+        return StartType.Warm
```
The else branch can just be delegated to `_start_type`

DanAlbert commented on gen/aircraft.py:
```diff
@@ -265,6 +266,25 @@ def _start_type(start_type: str) -> StartType:
+        unit_type: Type[FlyingType],
+        at: Union[ShipGroup, StaticGroup],
+    ) -> StartType:
+        group_units = at.dict().get("units")
```
This should be `at.units`, guarded with an `isinstance` if needed. We shouldn't be dependent on the serialization format of the miz.

MetalStormGhost committed:
Replaced at.dict().get("units") with at.units and removed redundant code with a call to _start_type

MetalStormGhost commented:
Replaced at.dict().get("units") with at.units as requested. I did not add an isinstance check as I couldn't see where that would be needed in this case.

Also removed the redundant code, now calling _start_type instead.

DanAlbert commented:
> I did not add an isinstance check as I couldn't see where that would be needed in this case.

I wasn't sure if it would be needed by mpyy (if `StaticGroup` didn't also have a `units` member). Looks like no, so not needed.
